### PR TITLE
[ENH] Series search for similarity search module

### LINKED ISSUE #1900: [ENH] Series search for similarity search module
### Describe the feature or idea you want to propose

Following #1243, we introduced query search, which was defined as : 

- Query search : Given a query Q and a series/collection X, evaluate the similarity between Q and each admissible candidate in X.

We now want to continue with the plan that was presented to implement series search, which was defined as  :

- Series search : Given a length parameter (for now, we add techniques that don't require it later) do a query search for all admissible queries in a series/collection X. In the naive case, this is simply a broadcasting of query search, but more optimized algorithms exists for this case (e.g. STUMP/STOMP for Euclidean distance)

### Describe your proposed solution

Implement it in the same design as query search (parameterize output and distance used in init, as well as speed-up) and by default interface it with `query_search` when no specific speed-up for `series_search` is available.

For example, use Mueen algorithm for Euclidean distance, if STAMP is not available. 

### Describe alternatives you've considered, if relevant

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by baraline**
#### Reference Issues/PRs

Fixes #1900 

#### What does this implement/fix? Explain your changes.

After the query search, this PR introduces the series search, which is linked to the task of computing matrix profiles. 
To make this task compatible with query search (for the naive approach), I changed the return of the predict function for query search to return both index and distances to best matches.
Modified docstrings accordingly. 

For now, only the naive version (the one looping on query search, which is equivalent to STAMP from matrix profile paper 1) and STOMP (from matrix profile paper 2) algorithm for squared and euclidean distance are implemented. I'll add the different matrix profiles (e.g. SWAMP with dtw and more efficient euclidean ones) methods in a later PR.

Minimal functionality were tested given time frame for ECML, more complete test are planned in next PR.

#### Comments

I'm still not entirely satisfied with the naming of query search / series search, so that's open to discussion

##### For all contributions
- [X] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you.
- [X] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

##### For new estimators and functions
- [x] I've added the estimator to the online [API documentation](https://www.aeon-toolkit.org/en/latest/api_reference.html).
- [X] (OPTIONAL) I've added myself as a `__maintainer__` at the top of relevant files and want to be contacted regarding its maintenance. Unmaintained files may be removed. This is for the full file, and you should not add yourself if you are just making minor changes or do not want to help maintain its contents.

##### For developers with write access
- [x] (OPTIONAL) I've updated aeon's [CODEOWNERS](https://github.com/aeon-toolkit/aeon/blob/main/CODEOWNERS) to receive notifications about future changes to these files.

### CODE REVIEW

**Comment by baraline** [comment]
Will update notebooks tomorrow and this will be good for review ! (and fix test that fails while I'll sleep)

**Comment by baraline** [comment]
Switching this back to draft as there is huge performance issue with STOMP

File: aeon/utils/validation/collection.py [hunk 1]
```diff
@@ -262,7 +264,7 @@ def get_type(X):
             raise ValueError(
                 f"ERROR np.ndarray must be 2D or 3D but found " f"{X.ndim}"
             )
-    elif isinstance(X, list):  # np-list or df-list
```
Comment (MatthewMiddlehurst) [inline aeon/utils/validation/collection.py:265]: Is this necessary? Needs more discussion IMO.

File: aeon/utils/numba/general.py [hunk 2]
```diff
@@ -36,7 +36,7 @@
 
 import aeon.utils.numba.stats as stats
 
-AEON_NUMBA_STD_THRESHOLD = 1e-8
+AEON_NUMBA_STD_THRESHOLD = 1e-12
```
Comment (MatthewMiddlehurst) [inline aeon/utils/numba/general.py:39]: Any reason why?

**Review by MatthewMiddlehurst — COMMENTED** [review]
Going to approve the module changes in this given the upcoming ECML tutorial, the fact the module is experimental and the next release after this is v1.0.0 with major changes.

We should have a more thorough review prior to v1.0.0 I feel, however.

Couple of comments for other changes though.

File: aeon/utils/validation/collection.py (hunk 1 above)
Comment (baraline) [reply aeon/utils/validation/collection.py:265]: Don't remember doing this one also ... I'll check if this is necessary

File: aeon/utils/numba/general.py (hunk 2 above)
Comment (baraline) [reply aeon/utils/numba/general.py:39]: Think it's an aterfact of when I add the issue with EEG dataset, which has a very small scale and thus naturally produces std values under 1e-8.
I'm fine with staying with 1e-8 for now, it is something that would need a more thorough inspection later as you said.

**Comment by baraline** [comment]
I agree that some changes would be needed before 1.0 to make thing cleaner, multiple issues are already up for this purpose.

**Review by MatthewMiddlehurst — APPROVED** [review]

**Pull request merged**